backend/src/main/java/com/fortunator/api/service/TransactionService.java

```diff
@@ -0,0 +1,57 @@
+package com.fortunator.api.service;
+
+import java.time.LocalDate;
+import java.util.List;
+import java.util.Optional;
+
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Service;
+
+import com.fortunator.api.models.Transaction;
+import com.fortunator.api.repository.TransactionRepository;
+import com.fortunator.api.service.exceptions.ResourceNotFoundException;
+import com.fortunator.api.service.exceptions.UserNotFoundException;
+
+@Service
+public class TransactionService {
+
+	@Autowired
+	private TransactionCategoryService transactionCategoryService;
+
+	@Autowired
+	private UserService userService;
+
+	@Autowired
+	private TransactionRepository transactionRepository;
+
+	@Autowired
+	private BalanceService balanceService;
+
+	public Transaction createTransaction(Transaction transaction) {
+		if (!transactionCategoryService.findById(transaction.getTransactionCategory().getId()).isPresent()) {
+			throw new ResourceNotFoundException("Category not found");
+		} else {
+			transaction.setTransactionCategory(
+					transactionCategoryService.findById(transaction.getTransactionCategory().getId()).get());
+		}
+		if (!userService.findUserById(transaction.getUser().getId()).isPresent()) {
+			throw new UserNotFoundException("User not found");
+		} else {
+			transaction.setUser(userService.findUserById(transaction.getUser().getId()).get());
+		}
+		balanceService.updateBalance(transaction.getUser().getId(), transaction.getAmount(),
```

Comment:
Alguns pontos aqui: Pensando em otimizar esse trecho de código, não precisamos fazer as mesmas queries serem executadas duas vezes. Estamos buscando, por exemplo, a `TransactionCategory` no banco para verificar se existe e aí, caso exista, buscamos de novo para setar dentro de transaction. Isso também acontece para o user. 

```suggestion
               Long categoryId = transaction.getTransactionCategory().getId();
               TransactionCategory category = transactionCategoryService.findById(categoryId);
		
		if (!category.isPresent()) {
			throw new ResourceNotFoundException("Category not found");
                }
		
		transaction.setTransactionCategory(category.get());
		
		User user = userService.findUserById(transaction.getUser().getId());
		
		if (!user.isPresent()) {
			throw new UserNotFoundException("User not found");
		}
		
		transaction.setUser(user.get());
		
		balanceService.updateBalance(transaction.getUser().getId(), transaction.getAmount(),
```

Provável que tenha algum error de sintaxe nessa sugestão, mas segue um possível caminho de como optimizar esse bloco de código  :grimacing:

Reply:
E por fim, outra sugestão de otimização que daria  mas talvez isso seja mais complexo em java rs :

Essas  duas queries, de usuário e categoria, poderiam acontecer em paralelo, já que uma não depende da outra. Isso daria um bom ganho de desempenho. Se for coisa simples de fazer, acho interessante incluir também, senão de boas hashua